securityscanutils/securityscan.go

```diff
@@ -0,0 +1,281 @@
+	   ' >= 1.7': ["gloo", "discovery", "rate-limit"]
+	*/
+	// where the patch number is explicitly not set so that these versions can match all
+	// 1.6.x-x releases
+	ImagesPerVersion map[string][]string
+	// VersionConstraint on releases to security scan
+	// any releases that do not pass this constraint will not be security scanned.
+	// If left empty, all versions will be scanned
+	VersionConstraint *semver.Constraints
+
+	// Required: image repo (quay.io, grc.io, gchr.io)
+	ImageRepo string
+
+	// Uploads Sarif file to github security code-scanning results
+	// e.g. https://github.com/solo-io/gloo/security/code-scanning
+	UploadCodeScanToGithub bool
+}
+
+// Main method to call on SecurityScanner which generates .md and .sarif files
+// in OutputDir as defined above per repo. If UploadCodeScanToGithub is true,
+// sarif files will be uploaded to the repository's code-scanning endpoint.
+func (s *SecurityScanner) GenerateSecurityScans(ctx context.Context) error {
+
+	client, err := githubutils.GetClient(ctx)
+	if err != nil {
+		return eris.Wrap(err, "error initializing github client")
+	}
+	markdownTplFile, err := GetTemplateFile(MarkdownTrivyTemplate)
+	if err != nil {
+		return eris.Wrap(err, "error creating temporary markdown template file to pass to trivy")
+	}
+	sarifTplFile, err := GetTemplateFile(SarifTrivyTemplate)
+	if err != nil {
+		return eris.Wrap(err, "error creating temporary markdown template file to pass to trivy")
+	}
+	defer func() {
+		os.Remove(markdownTplFile)
+		os.Remove(sarifTplFile)
+	}()
+	for _, repo := range s.Repos {
+		opts := repo.Opts
+		allReleases, err := githubutils.GetAllRepoReleases(ctx, client, repo.Owner, repo.Repo)
+		if err != nil {
+			return eris.Wrapf(err, "unable to fetch all github releases for github.com/%s/%s", repo.Owner, repo.Repo)
+		}
+		// Filter releases by version constraint provided
+		filteredReleases := githubutils.FilterReleases(allReleases, opts.VersionConstraint)
+		githubutils.SortReleasesBySemver(filteredReleases)
+		for _, release := range filteredReleases {
+			// We can swallow the error here, any releases with improper tag names
+			// will not be included in the filtered list
+			ver, _ := semver.NewVersion(release.GetTagName())
+			err = repo.RunMarkdownScan(ver, markdownTplFile)
+			if err != nil {
+				return eris.Wrapf(err, "error generating markdown file from security scan for version %s", release.GetTagName())
+			}
+			// Only generate sarif files if we are uploading code scan results
+			// to github
+			if repo.Opts.UploadCodeScanToGithub {
+				err = repo.RunGithubSarifScan(ver, sarifTplFile)
+				if err != nil {
+					return eris.Wrapf(err, "error generating github sarif file from security scan for version %s", release.GetTagName())
+				}
+			}
+		}
+
+	}
+	return nil
+}
+
+func (r *SecurityScanRepo) RunMarkdownScan(versionToScan *semver.Version, markdownTplFile string) error {
+	images, err := r.GetImagesToScan(versionToScan)
```

Comment:
Do we need to filter these images? I thought executing `filteredReleases := githubutils.FilterReleases(allReleases, opts.VersionConstraint)` which we do earlier would protect us against scanning images that don't fit the constraint

Reply:
FilterReleases uses the VersionConstraint defined [here](https://github.com/solo-io/go-utils/blob/108a37f922d9598980952134bf659310d78ff694/securityscanutils/securityscan.go#L66). This `GetImagesToScan` uses the [ImagesPerVersion](https://github.com/solo-io/go-utils/blob/108a37f922d9598980952134bf659310d78ff694/securityscanutils/securityscan.go#L62) map to figure out which images belong to which version. We are not filtering anything here, but rather retrieving which images actually belong to this verison.